In [35]:
%load_ext sql
%sql mysql+pymysql://root:rootroot@localhost:3306/Target

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


'Connected: root@Target'

### 1. Import the dataset and do usual exploratory analysis steps like checking the structure & characteristics of the dataset

#### 1.1 Data type of all columns in the "customers" table.

In [11]:
%%sql
Describe Target.customers

 * mysql+pymysql://root:***@localhost:3306/Target
5 rows affected.


[('customer_id', 'varchar(60)', 'YES', '', None, ''),
 ('customer_unique_id', 'varchar(60)', 'YES', '', None, ''),
 ('customer_zip_code_prefix', 'mediumint', 'YES', '', None, ''),
 ('customer_city', 'varchar(50)', 'YES', '', None, ''),
 ('customer_state', 'char(2)', 'YES', '', None, '')]

#### 1.2 Get the time range between which the orders were placed.

In [5]:
%%sql
SELECT MIN(order_purchase_timestamp) first_date, MAX(order_purchase_timestamp) last_date
from Target.orders;

 * mysql+pymysql://root:***@localhost:3306/Target
1 rows affected.


[(datetime.datetime(2016, 9, 4, 21, 15, 19), datetime.datetime(2018, 10, 17, 17, 30, 18))]

#### 1.3. Count the Cities & States of customers who ordered during the given period.

In [12]:
%%sql
SELECT
  count(DISTINCT c.customer_state) states,
  count(DISTINCT c.customer_city) cities
from
  Target.orders
  join Target.customers c  USING(customer_id)


 * mysql+pymysql://root:***@localhost:3306/Target
1 rows affected.


[(27, 4119)]

### 2. In-depth Exploration

#### 2.1. Is there a growing trend in the no. of orders placed over the past years?

In [13]:
%%sql
select
  EXTRACT( year from order_purchase_timestamp) AS year,
  count(*) order_count
from
  Target.orders
GROUP BY year
ORDER BY year


 * mysql+pymysql://root:***@localhost:3306/Target
3 rows affected.


[(2016, 329), (2017, 45101), (2018, 54011)]

In [14]:
%%sql
select
  EXTRACT(year from order_purchase_timestamp)  year,
  EXTRACT(month from order_purchase_timestamp) month,
  count(*) order_count
from
  Target.orders
GROUP by 
  year, month
order by 
  year, month


 * mysql+pymysql://root:***@localhost:3306/Target
25 rows affected.


[(2016, 9, 4),
 (2016, 10, 324),
 (2016, 12, 1),
 (2017, 1, 800),
 (2017, 2, 1780),
 (2017, 3, 2682),
 (2017, 4, 2404),
 (2017, 5, 3700),
 (2017, 6, 3245),
 (2017, 7, 4026),
 (2017, 8, 4331),
 (2017, 9, 4285),
 (2017, 10, 4631),
 (2017, 11, 7544),
 (2017, 12, 5673),
 (2018, 1, 7269),
 (2018, 2, 6728),
 (2018, 3, 7211),
 (2018, 4, 6939),
 (2018, 5, 6873),
 (2018, 6, 6167),
 (2018, 7, 6292),
 (2018, 8, 6512),
 (2018, 9, 16),
 (2018, 10, 4)]

**Judging from above data we can see that there was a increase in order count from 2016 to 2017, but the growth was minimal from 2017 to 2018**

#### 2.2 Can we see some kind of monthly seasonality in terms of the no. of orders being placed?

**No seasonality is observed. This may be due to the size of dataset**



#### 2.3. During what time of the day, do the Brazilian customers mostly place their orders? (Dawn, Morning, Afternoon or Night)
- 0-6 hrs : Dawn
- 7-12 hrs : Mornings
- 13-18 hrs : Afternoon
- 19-23 hrs : Night

In [15]:
%%sql
select case
           when extract( hour from order_purchase_timestamp) BETWEEN 0 and 6 THEN 'Dawn'
           when extract( hour from order_purchase_timestamp) BETWEEN 7 and 12 THEN 'Mornings'
           when extract( hour from order_purchase_timestamp) BETWEEN 13 and 18 THEN 'Afternoon'
           else 'Night'
       end as order_time,
    count(*) order_count
from Target.orders
GROUP by order_time
ORDER BY order_count desc

 * mysql+pymysql://root:***@localhost:3306/Target
4 rows affected.


[('Afternoon', 38135), ('Night', 28331), ('Mornings', 27733), ('Dawn', 5242)]

**Majority of customers order during the afternoon time**

### 3. Evolution of E-commerce orders in the Brazil region

#### 3.1 Get the month on month no. of orders placed in each state

In [17]:
%%sql
SELECT customer_state,
    EXTRACT(year from order_purchase_timestamp)  year,
    EXTRACT(month from order_purchase_timestamp) month,
    count(DISTINCT order_id) order_count
from Target.orders
    JOIN Target.customers  USING(customer_id) 
GROUP BY customer_state,
         year,month
ORDER BY customer_state,
         year,month
LIMIT 10

 * mysql+pymysql://root:***@localhost:3306/Target
10 rows affected.


[('AC', 2017, 1, 2),
 ('AC', 2017, 2, 3),
 ('AC', 2017, 3, 2),
 ('AC', 2017, 4, 5),
 ('AC', 2017, 5, 8),
 ('AC', 2017, 6, 4),
 ('AC', 2017, 7, 5),
 ('AC', 2017, 8, 4),
 ('AC', 2017, 9, 5),
 ('AC', 2017, 10, 6)]

#### 3.2. How are the customers distributed across all the states?

In [18]:
%%sql
SELECT customer_state,
    count(distinct customer_id) customer_count
from Target.customers
GROUP BY customer_state
ORDER BY customer_count desc

 * mysql+pymysql://root:***@localhost:3306/Target
27 rows affected.


[('SP', 41746),
 ('RJ', 12852),
 ('MG', 11635),
 ('RS', 5466),
 ('PR', 5045),
 ('SC', 3637),
 ('BA', 3380),
 ('DF', 2140),
 ('ES', 2033),
 ('GO', 2020),
 ('PE', 1652),
 ('CE', 1336),
 ('PA', 975),
 ('MT', 907),
 ('MA', 747),
 ('MS', 715),
 ('PB', 536),
 ('PI', 495),
 ('RN', 485),
 ('AL', 413),
 ('SE', 350),
 ('TO', 280),
 ('RO', 253),
 ('AM', 148),
 ('AC', 81),
 ('AP', 68),
 ('RR', 46)]

**From above data we can infer that majority of the customer are from Sao Paulo**

### 4. Impact on Economy: Analyze the money movement by e-commerce by looking at order prices, freight and others. 

#### 4.1. Get the % increase in the cost of orders from year 2017 to 2018 (include months between Jan to Aug only). You can use the "payment_value" column in the payments table to get the cost of orders.

In [19]:
%%sql
with
  cte
  as
  (
    SELECT
      EXTRACT(month FROM order_purchase_timestamp) month,
      SUM(case when EXTRACT(YEAR FROM order_purchase_timestamp) = 2017 AND EXTRACT(month FROM order_purchase_timestamp) BETWEEN 2 AND 7 then payment_value else 0 end) data_17,
      SUM(case when EXTRACT(YEAR FROM order_purchase_timestamp) = 2018 AND EXTRACT(month FROM order_purchase_timestamp) BETWEEN 2 AND 7 then payment_value else 0 end) data_18
    FROM target.payments
      JOIN target.orders
  USING (order_id) 
    WHERE 
    EXTRACT(YEAR FROM order_purchase_timestamp) BETWEEN 2017 AND 2018 AND EXTRACT(month FROM order_purchase_timestamp) BETWEEN 2 AND 7
    GROUP BY month
    ORDER BY month
  )
SELECT
  month,
  round(((data_18 - data_17)/ data_17)*100,2) pct_chg
from cte 

 * mysql+pymysql://root:***@localhost:3306/Target
6 rows affected.


[(2, 239.99), (3, 157.78), (4, 177.84), (5, 94.63), (6, 100.26), (7, 80.04)]

**From above data we can say that February month showed highest change in order cost while July showed the lowest**

#### 4.2. Calculate the Total & Average value of order price for each state.

In [20]:
%%sql
SELECT customer_state,
    round(sum(payment_value)) total_value,
    round(avg(payment_value), 2) average_value
from Target.payments
    JOIN Target.orders USING(order_id)
    JOIN Target.customers USING(customer_id)  
GROUP BY customer_state
ORDER BY average_value DESC

 * mysql+pymysql://root:***@localhost:3306/Target
27 rows affected.


[('PB', 141546.0, 248.33),
 ('AC', 19681.0, 234.29),
 ('RO', 60866.0, 233.2),
 ('AP', 16263.0, 232.33),
 ('AL', 96962.0, 227.08),
 ('RR', 10065.0, 218.8),
 ('PA', 218296.0, 215.92),
 ('SE', 75246.0, 208.44),
 ('PI', 108524.0, 207.11),
 ('TO', 61485.0, 204.27),
 ('CE', 279464.0, 199.9),
 ('MA', 152523.0, 198.86),
 ('RN', 102718.0, 196.78),
 ('MT', 187029.0, 195.23),
 ('PE', 324850.0, 187.99),
 ('MS', 137535.0, 186.87),
 ('AM', 27967.0, 181.6),
 ('BA', 616646.0, 170.82),
 ('SC', 623086.0, 165.98),
 ('GO', 350092.0, 165.76),
 ('DF', 355141.0, 161.13),
 ('RJ', 2144380.0, 158.53),
 ('RS', 890899.0, 157.18),
 ('MG', 1872257.0, 154.71),
 ('ES', 325968.0, 154.71),
 ('PR', 811156.0, 154.15),
 ('SP', 5998227.0, 137.5)]

**From above data we can say that max average order value is 248.44 originating from PB state**

#### 4.3. Calculate the Total & Average value of order freight for each state.

In [21]:
%%sql
select customer_state,
    round(sum(freight_value)) total_freight_value,
    round(avg(freight_value),2) avg_freight_value
from Target.orders
    join Target.order_items using(order_id)
    join Target.customers using(customer_id)  
group by customer_state
ORDER BY avg_freight_value DESC

 * mysql+pymysql://root:***@localhost:3306/Target
27 rows affected.


[('RR', 2235.0, 42.98),
 ('PB', 25720.0, 42.72),
 ('RO', 11417.0, 41.07),
 ('AC', 3687.0, 40.07),
 ('PI', 21218.0, 39.15),
 ('MA', 31524.0, 38.26),
 ('TO', 11733.0, 37.25),
 ('SE', 14111.0, 36.65),
 ('AL', 15915.0, 35.84),
 ('PA', 38699.0, 35.83),
 ('RN', 18860.0, 35.65),
 ('AP', 2789.0, 34.01),
 ('AM', 5479.0, 33.21),
 ('PE', 59450.0, 32.92),
 ('CE', 48352.0, 32.71),
 ('MT', 29715.0, 28.17),
 ('BA', 100157.0, 26.36),
 ('MS', 19144.0, 23.37),
 ('GO', 53115.0, 22.77),
 ('ES', 49765.0, 22.06),
 ('RS', 135523.0, 21.74),
 ('SC', 89660.0, 21.47),
 ('DF', 50626.0, 21.04),
 ('RJ', 305589.0, 20.96),
 ('MG', 270853.0, 20.63),
 ('PR', 117852.0, 20.53),
 ('SP', 718723.0, 15.15)]

**From above data we can say that max average freight value is 42.98 originating from RR state**

### 5. Analysis based on sales, freight and delivery time

#### 5.1 Find the no. of days taken to deliver each order from the order’s purchase date as delivery time. Also, calculate the difference (in days) between the estimated & actual delivery date of an order.Do this in a single query.

In [23]:
%%sql
SELECT order_id,
    DATEDIFF(order_delivered_customer_date, order_purchase_timestamp) as time_to_deliver,
    DATEDIFF( order_estimated_delivery_date,  order_delivered_customer_date) as diff_estimated_delivery
from Target.orders
where
  DATEDIFF(order_delivered_customer_date, order_purchase_timestamp) IS NOT NULL
ORDER by time_to_deliver desc
LIMIT 10

 * mysql+pymysql://root:***@localhost:3306/Target
10 rows affected.


[('ca07593549f1816d26a572e06dc1eab6', 210, -181),
 ('1b3190b2dfa9d789e1f14c05b647a14a', 208, -188),
 ('440d0d17af552815d15a9e41abe49359', 196, -165),
 ('2fb597c2f772eca01b1f5c561bf6cc7b', 195, -155),
 ('285ab9426d6982034523a855f55a885e', 195, -166),
 ('0f4519c5f1c541ddec9f21b3bddd533a', 194, -161),
 ('47b40429ed8cce3aee9199792275433f', 191, -175),
 ('2fe324febf907e3ea3f2aa9650869fa5', 190, -167),
 ('2d7561026d542c8dbd8f0daeadf67a43', 188, -159),
 ('c27815f7e3dd0b926b58552628481575', 188, -162)]

#### 5.2.a Find out the top 5 states with the highest freight value

In [24]:
%%sql
SELECT customer_state Top_5,
    round(avg(freight_value), 2) as avg_freight_value
from Target.orders
    join Target.order_items using(order_id)
    join Target.customers using(customer_id)  
GROUP BY customer_state
order by avg_freight_value desc
limit 5

 * mysql+pymysql://root:***@localhost:3306/Target
5 rows affected.


[('RR', 42.98), ('PB', 42.72), ('RO', 41.07), ('AC', 40.07), ('PI', 39.15)]

#### 5.2.b Find out the top 5 states with the  lowest average freight value

In [25]:
%%sql
SELECT customer_state Bottom_5,
    round(avg(freight_value), 2) as avg_freight_value
from Target.orders
    join Target.order_items using(order_id)
    join Target.customers using(customer_id)  
GROUP BY customer_state
order by avg_freight_value
limit 5

 * mysql+pymysql://root:***@localhost:3306/Target
5 rows affected.


[('SP', 15.15), ('PR', 20.53), ('MG', 20.63), ('RJ', 20.96), ('DF', 21.04)]

#### 5.3.a Find out the top 5 states with the highest delivery time.

In [27]:
%%sql
SELECT customer_state Top_5,
    avg(DATEDIFF(order_delivered_customer_date, order_purchase_timestamp)) as avg_delivery_time
from Target.orders
    join Target.customers using(customer_id) 
GROUP BY customer_state
order by avg_delivery_time DESC
limit 5

 * mysql+pymysql://root:***@localhost:3306/Target
5 rows affected.


[('RR', Decimal('29.3415')),
 ('AP', Decimal('27.1791')),
 ('AM', Decimal('26.3586')),
 ('AL', Decimal('24.5013')),
 ('PA', Decimal('23.7252'))]

#### 5.3.b Find out the top 5 states with the lowest average delivery time.

In [29]:
%%sql
SELECT customer_state Top_5,
    avg(DATEDIFF(order_delivered_customer_date, order_purchase_timestamp)) as avg_delivery_time
from target.orders
    join Target.customers using(customer_id) 
GROUP BY customer_state
order by avg_delivery_time
limit 5

 * mysql+pymysql://root:***@localhost:3306/Target
5 rows affected.


[('SP', Decimal('8.7005')),
 ('PR', Decimal('11.9380')),
 ('MG', Decimal('11.9465')),
 ('DF', Decimal('12.8990')),
 ('SC', Decimal('14.9075'))]

#### 5.4 Find out the top 5 states where the order delivery is really fast as compared to the estimated date of delivery.

In [30]:
%%sql
SELECT
    customer_state,
    avg(DATEDIFF(order_estimated_delivery_date, order_purchase_timestamp) -  DATEDIFF(order_delivered_customer_date, order_purchase_timestamp)) diff
FROM target.orders
    join Target.customers using(customer_id) 
GROUP BY customer_state
ORDER BY diff
limit 5

 * mysql+pymysql://root:***@localhost:3306/Target
5 rows affected.


[('AL', Decimal('8.7078')),
 ('MA', Decimal('9.5718')),
 ('SE', Decimal('10.0209')),
 ('ES', Decimal('10.4962')),
 ('BA', Decimal('10.7945'))]

### 6. Analysis based on the payments

#### 6.1 Find the month on month no. of orders placed using different payment types.

In [9]:
SELECT DISTINCT payment_type from Target.payments

(5 row(s) affected)

Total execution time: 00:00:01.098

payment_type credit_card UPI voucher debit_card not_defined

In [31]:
%%sql
SELECT
  EXTRACT(year from order_purchase_timestamp)  year,
  EXTRACT(month from order_purchase_timestamp) month,
  sum(case when payment_type = "credit_card" then 1 else 0 end) credit_card,
  sum(case when payment_type = "UPI" then 1 else 0 end) UPI,
  sum(case when payment_type = "voucher" then 1 else 0 end) voucher,
  sum(case when payment_type = "debit_card" then 1 else 0 end) debit_card,
  sum(case when payment_type = "not_defined" then 1 else 0 end) not_defined
from
  Target.orders
  join Target.payments using(order_id) 
GROUP BY 
  year,month
ORDER BY 
  year,month

 * mysql+pymysql://root:***@localhost:3306/Target
25 rows affected.


[(2016, 9, Decimal('3'), Decimal('0'), Decimal('0'), Decimal('0'), Decimal('0')),
 (2016, 10, Decimal('254'), Decimal('63'), Decimal('23'), Decimal('2'), Decimal('0')),
 (2016, 12, Decimal('1'), Decimal('0'), Decimal('0'), Decimal('0'), Decimal('0')),
 (2017, 1, Decimal('583'), Decimal('197'), Decimal('61'), Decimal('9'), Decimal('0')),
 (2017, 2, Decimal('1356'), Decimal('398'), Decimal('119'), Decimal('13'), Decimal('0')),
 (2017, 3, Decimal('2016'), Decimal('590'), Decimal('200'), Decimal('31'), Decimal('0')),
 (2017, 4, Decimal('1846'), Decimal('496'), Decimal('202'), Decimal('27'), Decimal('0')),
 (2017, 5, Decimal('2853'), Decimal('772'), Decimal('289'), Decimal('30'), Decimal('0')),
 (2017, 6, Decimal('2463'), Decimal('707'), Decimal('239'), Decimal('27'), Decimal('0')),
 (2017, 7, Decimal('3086'), Decimal('845'), Decimal('364'), Decimal('22'), Decimal('0')),
 (2017, 8, Decimal('3284'), Decimal('938'), Decimal('294'), Decimal('34'), Decimal('0')),
 (2017, 9, Decimal('3283'), Decimal('903'), Decimal('287'), Decimal('43'), Decimal('0')),
 (2017, 10, Decimal('3524'), Decimal('993'), Decimal('291'), Decimal('52'), Decimal('0')),
 (2017, 11, Decimal('5897'), Decimal('1509'), Decimal('387'), Decimal('70'), Decimal('0')),
 (2017, 12, Decimal('4377'), Decimal('1160'), Decimal('294'), Decimal('64'), Decimal('0')),
 (2018, 1, Decimal('5520'), Decimal('1518'), Decimal('416'), Decimal('109'), Decimal('0')),
 (2018, 2, Decimal('5253'), Decimal('1325'), Decimal('305'), Decimal('69'), Decimal('0')),
 (2018, 3, Decimal('5691'), Decimal('1352'), Decimal('391'), Decimal('78'), Decimal('0')),
 (2018, 4, Decimal('5455'), Decimal('1287'), Decimal('370'), Decimal('97'), Decimal('0')),
 (2018, 5, Decimal('5497'), Decimal('1263'), Decimal('324'), Decimal('51'), Decimal('0')),
 (2018, 6, Decimal('4813'), Decimal('1100'), Decimal('324'), Decimal('182'), Decimal('0')),
 (2018, 7, Decimal('4755'), Decimal('1229'), Decimal('281'), Decimal('242'), Decimal('0')),
 (2018, 8, Decimal('4985'), Decimal('1139'), Decimal('295'), Decimal('277'), Decimal('2')),
 (2018, 9, Decimal('0'), Decimal('0'), Decimal('15'), Decimal('0'), Decimal('1')),
 (2018, 10, Decimal('0'), Decimal('0'), Decimal('4'), Decimal('0'), Decimal('0'))]

**From above data we can infer that majority of users use credit card for purchasing**

#### 6.2 Find the no. of orders placed on the basis of the payment installments that have been paid.

In [32]:
%%sql
SELECT
  payment_installments, COUNT(order_id) AS order_count
from Target.payments
  join Target.orders using(order_id) 
where order_status != 'canceled'
GROUP BY payment_installments
ORDER BY order_count DESC;

 * mysql+pymysql://root:***@localhost:3306/Target
24 rows affected.


[(1, 52184),
 (2, 12353),
 (3, 10392),
 (4, 7056),
 (10, 5292),
 (5, 5209),
 (8, 4239),
 (6, 3898),
 (7, 1620),
 (9, 638),
 (12, 133),
 (15, 74),
 (18, 27),
 (11, 22),
 (24, 18),
 (20, 17),
 (13, 15),
 (14, 15),
 (17, 8),
 (16, 5),
 (21, 3),
 (0, 2),
 (22, 1),
 (23, 1)]

**From above data we can infer that majority of the payments are done as single installments and max installment is 23**